### PR TITLE
[Structured Output] Let terminal grammars stop under min_tokens (II)

### PR DESCRIPTION
**Pull request by arpera**
## Purpose

min_tokens can force structured-output requests to sample outside the grammar after a short schema reaches its terminal state. For schemas such as {"type":"boolean"}, {"enum":["a"]}, or {"type":"null"}, the grammar's only legal continuation after the value is EOS. min_tokens then masks EOS because the request has not reached the requested minimum length, leaving no legal token.

This PR was created as a continuation of https://github.com/vllm-project/vllm/pull/47489 and https://github.com/vllm-project/vllm/pull/49521 to fix the issue reported also here https://github.com/vllm-project/vllm/issues/53665.
Also this PR fixes https://github.com/vllm-project/vllm/pull/51299 since the fix is in the same code area.

Moreover the issue with min_tokens + structured-output constraints is even worse. It can also be reproduced with any "semantic masking" logit processor like structured outputs, allowed_token_ids, min_tokens, bad_words that can mask out all token_ids (all logits are -inf). See original [comment](https://github.com/vllm-project/vllm/pull/47489#issuecomment-5062309103). This PR does not try to fix all the other logit processors, only focusing for now on min_tokens + structured output. Full fix requires larger refactor and is left for future attempts.

## Test Plan

<details>
<summary>Repro for PR #47489 for both MRV1 and MRV2 ([original gist](https://gist.github.com/yzong-rh/40200cc3831c38f9e79b44182328c08b))</summary>

```python
#!/usr/bin/env python3
"""Repro for min_tokens + terminal structured-output schemas (#47489)."""

from __future__ import annotations

import json
import os

from vllm import LLM, SamplingParams
from vllm.distributed.parallel_state import cleanup_dist_env_and_memory
from vllm.sampling_params import StructuredOutputsParams


def run(llm: LLM, label: str, schema: dict, min_tokens: int) -> None:
    params = SamplingParams(
        temperature=0.0,
        min_tokens=min_tokens,
        max_tokens=40,
        structured_outputs=StructuredOutputsParams(json=json.dumps(schema)),
    )
    output = llm.generate(["Answer:"], params, use_tqdm=False)[0].outputs[0]
    print(
        label,
        repr(output.text),
        len(output.token_ids),
        output.token_ids,
        output.finish_reason,
        output.stop_reason,
    )


def run_cases(llm: LLM, runner_label: str) -> None:
    cases = [
        ("control-bool", {"type": "boolean"}, 0),
        ("min-bool", {"type": "boolean"}, 20),
        ("min-enum", {"enum": ["a"]}, 20),
        ("min-null", {"type": "null"}, 30),
    ]
    for name, schema, min_tokens in cases:
        run(llm, f"{runner_label}/{name}", schema, min_tokens)


def main() -> None:
    for use_v2 in (False, True):
        runner_label = "v2" if use_v2 else "v1"
        os.environ["VLLM_USE_V2_MODEL_RUNNER"] = "1" if use_v2 else "0"
        print(f"=== Model runner {runner_label.upper()} ===")
        llm = LLM(
            model="Qwen/Qwen3-0.6B",
            enforce_eager=True,
            gpu_memory_utilization=0.30,
            max_model_len=512,
            structured_outputs_config={"backend": "xgrammar"},
        )
        try:
            run_cases(llm, runner_label)
        finally:
            del llm
            cleanup_dist_env_and_memory()


if __name__ == "__main__":
    main()
```
</details>

**PASSED**

<details>
<summary>Repro script for PR #51299</summary>

```python
from vllm import LLM, SamplingParams

llm = LLM(model="facebook/opt-125m", max_model_len=128, enforce_eager=True)
# hangs forever before this PR
llm.generate(["Hello, my name is"],
             SamplingParams(max_tokens=500, min_tokens=500, ignore_eos=True))
```
</details>

**PASSED**

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Review by yzong-rh — COMMENTED** [review]
Had a quick look and the changes look good. Thanks!
Lmk if you need a full review after you fix the CI issue.

**Review by abmfy — APPROVED** [review]
LGTM!

**Review by yzong-rh — APPROVED** [review]
LGTM. Good catch on the `tl.debug_barrier()`.

**Review by vadiklyutiy — APPROVED** [review]

**Comment by njhill** [comment]
Thanks @arpera! As alluded to in your PR description, I feel like we should have a more general solution to this. Hopefully we can come up with something clean :)

**Comment by arpera** [comment]
Yes, true, a more general solution is needed here for this kind of constraints conflict. I can make a separate issue report to list the things we need to think about.

**Comment by yzong-rh** [comment]
Copied from a slack thread (details may be outdated, gist should still be right):

I think there are 2 classes of logits processors
  - C1: Semantic masks (e.g. structured outputs, allowed_token_ids, min_tokens, bad_words)
  - C2: Distribution shapers (e.g. logits_bias, penalties, temperature, min_p, tok_k, top_p)

C1 can mask out all possible token_ids. https://github.com/vllm-project/vllm/pull/47489 fixes a specific case (min_tokens + structured ouputs), but it can happen with other combinations as well (e.g. allowed_token_ids + structured outputs).

**Ideally, we run all C1 first. Do a check to ensure some token can still be sampled. Then run C2.**

There seems to be a parity issue with MRV1 vs V2 as well
 
In MRV1:
```
1. [C1] grammar bitmask          # outside sampler
2. [C1] allowed_token_ids
3. [C1] bad_words
4. [C1] min_tokens               # via non_argmax_invariant LPs
5. [C2] logit_bias               # same LP list, after min_tokens
6. [C2] penalties
7. [C1] thinking_budget          # after penalties
8. greedy argmax (if needed)
9. [C2] temperature
10.[C2] min_p (+ other argmax-invariant LPs)
11.[C2] top_k / top_p
12. multinomial sample
```

MRV2:
```
1. [C1] grammar bitmask          # outside sampler
2. fused LogitBiasState kernel:
     [C1] allowed_token_ids
     [C2] logit_bias             # ← before min_tokens
     [C1] min_tokens
3. [C2] penalties
4. [C1] bad_words                # ← after penalties
5. [C2] temperature
6. [C2] min_p
7. [C2] top_k / top_p            # skipped inside apply when FlashInfer samples
8. Gumbel / FlashInfer sample
```
